### PR TITLE
feat: Support for device code flow on SIW v1

### PR DESCRIPTION
**Pull request by nakuljoshi-okta**
## Description:

Implement support for device code flow on legacy SIW v1. 
This feature is already implemented on OIE. Here is the PR https://github.com/okta/okta-signin-widget/pull/1816

Also changing idx remediation form name to `user-code`.

## PR Checklist

- [x] Have you verified the basic functionality for this change?

- [x] Added unit tests?

- [x] Added e2e tests

- [ ] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?

### Reviewers:
@haishengwu-okta 

### Issue:

- [OKTA-393191](https://oktainc.atlassian.net/browse/OKTA-393191)
- [OKTA-393194](https://oktainc.atlassian.net/browse/OKTA-393194)

### CODE REVIEW

File: src/LoginRouter.js [hunk 1]
```diff
@@ -135,6 +137,8 @@ export default BaseLoginRouter.extend({
     'signin/consent': 'consentRequired',
     'signin/admin-consent': 'adminConsentRequired',
     'signin/enroll-user': 'enrollUser',
+    'signin/device-activate-complete': 'deviceActivateComplete',
+    'activate': 'deviceActivate',
```
[the remark this answers is not in the document]
Comment (larsjohansen-okta) [reply src/LoginRouter.js:141]: how about `/signin/device/activate`?

[the remark this answers is not in the document]
Comment (nakuljoshi-okta) [reply src/LoginRouter.js:141]: I would rather do `/signin/device-activate` to keep it consistent with other url patterns here. One thing to note though, the flow is going to start with `/activate` and then redirect to `/signin/device-activate`. @larsjohansen-okta  we good with that?

File: src/DeviceActivateController.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  className: 'device-code-activate',
+  Model: {
+    props: {
+      userCode: ['string', true],
+      stateToken: 'string',
+    },
+    save: function() {
+      const self = this;
+      return this.doTransaction(function(transaction) {
+        return transaction.deviceActivate({
```
[the remark this answers is not in the document]
Comment (nakuljoshi-okta) [reply src/DeviceActivateController.js:29]: yup. Check [here](https://github.com/okta/okta-signin-widget/pull/1942/files#diff-e46245e44657d71d63d226065c17cd418b48930ba38b2fc972435124744beba0R9) which maps this function name to href.

File: src/DeviceActivateController.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    },
+  },
+  Form: {
+    noCancelButton: true,
+    autoSave: true,
+    save: function() {
+      return loc('oform.next', 'login');
+    },
+    title: function() {
+      return loc('oie.device.code.activate.title', 'login');
```
[the remark this answers is not in the document]
Comment (nakuljoshi-okta) [reply src/DeviceActivateController.js]: I removed `oie` prefix. There are few keys with `idx` prefix which are coming from backend so hard to change now. Should I just add new keys for legacy? they will be duplicates though.

File: test/testcafe/spec/DeviceCodeActivateLegacy_spec.js [hunk 4]
```diff
@@ -0,0 +1,182 @@
+import {RequestMock, RequestLogger, ClientFunction} from 'testcafe';
```
[the remark this answers is not in the document]
Comment (nakuljoshi-okta) [reply test/testcafe/spec/DeviceCodeActivateLegacy_spec.js]: done

File: assets/sass/v2/_device-code-terminal.scss [hunk 5]
```diff
@@ -8,4 +8,7 @@
     display: inline-block;
     width: 21px;
   }
+  .legacy-try-again-button {
```
[the remark this answers is not in the document]
Comment (nakuljoshi-okta) [reply assets/sass/v2/_device-code-terminal.scss]: done

File: src/DeviceActivateController.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (haishengwu-okta) [reply src/DeviceActivateController.js]: Use same key as long as possible.

> with idx prefix which are coming from backend

Are those keys used at both v1 and v2 flows?

File: src/LoginRouter.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (haishengwu-okta) [reply src/LoginRouter.js]: > the flow is going to start with /activate and then redirect to /signin/device-activate.

why need this redirect instead of going to `signin/device-activate` directly?

File: src/DeviceActivateTerminalController.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        return loc('idx.device.not.activated.internal.error', 'login');
+      }
+    },
+    formChildren: function() {
+      return [
+        FormType.View({
+          View: View.extend({
+            template: hbs(
+              '{{#if isDeviceCodeError}}\
+                  <a href="/activate" class="button button-primary text-align-c retry-button" data-se="try-again">\
```
Comment (haishengwu-okta) [inline src/DeviceActivateTerminalController.js:70]: `/signin/activate` ?

File: test/testcafe/framework/page-objects/DeviceCodeActivatePageObject.js [hunk 7]
```diff
@@ -71,4 +71,16 @@ export default class DeviceCodeActivatePageObject extends BasePageObject {
   isTryAgainButtonPresent() {
     return Selector('[data-se="try-again"]').exists;
   }
+
+  fillLegacyUserNameField(value) {
```
Comment (haishengwu-okta) [inline test/testcafe/framework/page-objects/DeviceCodeActivatePageObject.js:75]: make a separated file for v1. could live in `page-objects-v1`

File: test/testcafe/spec/v1/DeviceCodeActivate_spec.js [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+  await t.expect(deviceCodeActivatePageObject.getPageSubtitle()).eql('Follow the instructions on your device for next steps');
+  await t.expect(deviceCodeActivatePageObject.isTerminalSuccessIconPresent()).eql(true);
+  await t.expect(deviceCodeActivatePageObject.isBeaconTerminalPresent()).eql(false);
+  await t.expect(deviceCodeActivatePageObject.isTryAgainButtonPresent()).eql(false);
+});
+
+test.requestHooks(legacyDeviceCodeConsentDeniedMock)('should be able to get device not activated screen when consent is denied on legacy SIW', async t => {
+  const deviceCodeActivatePageObject = await setup(t);
+  await rerenderWidget({
+    stateToken: '00-dummy-state-token', //start with 00 to render legacy sign in widget
```
Comment (haishengwu-okta) [inline test/testcafe/spec/v1/DeviceCodeActivate_spec.js:115]: nit:
ideally, we could think of creating a whole new suite for running v1 testcafe tests.
e.g. `test:testcafe-ci-v1` which create widgetrc with v1 token and run specs only from `specs/v1` etc.

It's not really related to your changes but if you're interested, certainly appreciate to tackle it in separate PR.

File: src/LoginRouter.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (nakuljoshi-okta) [reply src/LoginRouter.js]: PM wants starting url to be `/activate` since its going to be entered by users on their phones so it needs to be easy to type in. So the flow is going to be 

1. user goes to foo.okta.com/activate
2. backend sets the state token and return login page
3. siw calls /introspect
4. /introspect returns `DEVICE_ACTIVATE` status which triggers route change to `signin/device-activate` and renders user code form. 

cc: @larsjohansen-okta

File: src/DeviceActivateTerminalController.js (hunk 6 above)
Comment (nakuljoshi-okta) [reply src/DeviceActivateTerminalController.js:70]: we want to start the flow from scratch, so /activate will set a new state token and call /introspect.

File: src/DeviceActivateController.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (nakuljoshi-okta) [reply src/DeviceActivateController.js]: ideally we would like to use the same keys for both v1 and v2 flow.

@iancorcoran-okta  would it be okay to remove `idx` prefix from keys [here](https://github.com/okta/okta-core/blob/6fc7e9ac4351eae63c25aa3f3cf414d4c3a7b383/components/platformx/idx/api/src/main/java/com/okta/api/mediation/services/auth/TerminalStateTokenHelper.java#L37), so that we can use the same keys on OIE and legacy sign in widget?  cc: @larsjohansen-okta

File: test/testcafe/framework/page-objects/DeviceCodeActivatePageObject.js (hunk 7 above)
Comment (nakuljoshi-okta) [reply test/testcafe/framework/page-objects/DeviceCodeActivatePageObject.js]: done

File: packages/@okta/i18n/src/properties/login_ok_PL.properties [hunk 9]
```diff
@@ -764,12 +764,12 @@ oie.consent.enduser.accept.label = 》Ýéš، îţ´š ɱé ⾼ئ䀕ヸ€홝
 oie.consent.enduser.deny.label = 》Ñö، îţ´š ñöţ ɱé ฐโ⾼ئ䀕ヸ€홝한Ӝฐโ《
 oie.consent.enduser.email.allow.description = 》Ţö çöñţîñûé، ŕéţûŕñ ţö ţĥé ðéṽîçé öŕ ŵîñðöŵ ŵĥéŕé ýöû ŕéǫûéšţéð ţĥé éɱåîļ ļîñķ· Ýöû ɱåý çļöšé ţĥîš ŵîñðöŵ åţ åñý ţîɱé· 한Ӝฐโ⾼ئ䀕ヸ€홝한Ӝฐโ⾼ئ䀕ヸ€홝한Ӝฐโ《
 oie.consent.enduser.deny.description = 》Ýöû ɱåý çļöšé ţĥîš ŵîñðöŵ åţ åñý ţîɱé· 홝한Ӝฐโ⾼ئ䀕ヸ€홝한Ӝฐโ《
-oie.return.to.original.tab = 》Þļéåšé ŕéţûŕñ ţö öŕîĝîñåļ ţåƀ· 䀕ヸ€홝한Ӝฐโ⾼ئ䀕ヸ€홝한Ӝฐโ《
-oie.device.code.activate.title = 》Åçţîṽåţé ýöûŕ ðéṽîçé €홝한Ӝฐโ⾼ئ䀕ヸ€홝한Ӝฐโ《
-oie.device.code.activate.subtitle = 》Ƒöļļöŵ ţĥé îñšţŕûçţîöñš öñ ýöûŕ ðéṽîçé ţö ĝéţ åñ åçţîṽåţîöñ çöðé ئ䀕ヸ€홝한Ӝฐโ⾼ئ䀕ヸ€홝한Ӝฐโ《
-oie.device.code.activate.label = 》Åçţîṽåţîöñ Çöðé ฐโ⾼ئ䀕ヸ€홝한Ӝฐโ《
-oie.device.code.activated.success.title = 》Ðéṽîçé åçţîṽåţéð Ӝฐโ⾼ئ䀕ヸ€홝한Ӝฐโ《
-oie.device.code.activated.error.title = 》Ðéṽîçé ñöţ åçţîṽåţéð €홝한Ӝฐโ⾼ئ䀕ヸ€홝한Ӝฐโ《
+oie.return.to.original.tab = 》Þļéåšé ŕéţûŕñ ţö ţĥé öŕîĝîñåļ ţåƀ· 한Ӝฐโ⾼ئ䀕ヸ€홝한Ӝฐโ《
```
Comment (larsjohansen-okta) [inline packages/@okta/i18n/src/properties/login_ok_PL.properties:767]: is this change intentional?

File: src/DeviceActivateController.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (nakuljoshi-okta) [reply src/DeviceActivateController.js]: > Use same key as long as possible.
> 
> > with idx prefix which are coming from backend
> 
> Are those keys used at both v1 and v2 flows?

@haishengwu-okta  @larsjohansen-okta  After looking up more v1 api response does not return any keys like v2 does. So would it be okay define v1 keys on siw side? they will same value as idx keys but would be better to separate them out so there is no confusion. We can follow up later with IDX team to check if we can use the same keys and remove `idx` prefix in future.

File: packages/@okta/i18n/src/properties/login_ok_PL.properties (hunk 9 above)
Comment (nakuljoshi-okta) [reply packages/@okta/i18n/src/properties/login_ok_PL.properties:767]: I ran the tool to generate this, not sure why it changed this line. @haishengwu-okta any thoughts?

Comment (nakuljoshi-okta) [reply packages/@okta/i18n/src/properties/login_ok_PL.properties]: good catch, fixed it.

File: src/DeviceActivateController.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (iancorcoran-okta) [reply src/DeviceActivateController.js]: That `TerminalStateTokenHelper` is only used by OIE

**Review by haishengwu-okta — APPROVED** [review]
